Preliminary Sddc Groups Implementation

dimitarproynov opened the pull request:
Due to lack of SDK support for SDDC Group APIs a low-level HTTP client had to be implemented to call the mentioned APIs.

This commit implements only Create SDDC Group, Add/Remove SDDCs in a Group, and Delete SDDC Group.

Create and Delete external TGW is to follow up in a separate commit.

Testing done:
make build
make test
golangci-lint run

Added a new E2E test for SDDC Groups
make testacc TESTARGS="-run=TestAccResourceSddcGroupZerocloud" - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

dimitarproynov committed:
Merge pull request #8 from vmware/main

Merge new version of release.yaml

Commit:
[Preliminary] Implement SDDC Group support

Due to lack of SDK support for SDDC Group APIs a low-level HTTP client had to be implemented to call the mentioned APIs.

This commit implements only Create SDDC Group, Add/Remove SDDCs in a Group, and Delete SDDC Group.

Create and Delete external TGW is to follow up in a separate commit.

Testing done:
make build
make test
golangci-lint run

Added a new E2E test for SDDC Groups
make testacc TESTARGS="-run=TestAccResourceSddcGroupZerocloud" - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

karaatanassov commented on vmc/constants/constants.go:
```diff
@@ -68,7 +68,7 @@ const (
 	CspUrl         string = "CSP_URL"
 	ApiToken       string = "API_TOKEN"
-	OrgID          string = "ORG_ID"
+	OrgId          string = "ORG_ID"
```
revive reports this a linter error

`struct field OrgId should be OrgIDgo-revive`

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

karaatanassov replied:
It strikes me that all the *Url things are wrong too. Doe snot this code go through linter like Revive?

dimitarproynov replied:
Added the revive linter to the golangci run and fixed the issues it pointed out.

karaatanassov commented on vmc/task/task_v2_client.go:
```diff
@@ -0,0 +1,126 @@
+		cspUrl:       cspUrl,
+		refreshToken: refreshToken,
+		orgId:        orgId,
+		httpClient:   http.DefaultClient,
```
could this be issue if other code in the same process changes settings for this default client?

Also this precludes configuration of settings, right?

dimitarproynov replied:
The terraform cli calls the declared lifecycle methods ot the provider binary via grpc, so the provider is actually a different process.  There is no code that changes settings of the default http client compiled into the terraform provider binary, so there is no issue here.

karaatanassov commented on vmc/task/task_v2_client.go:
```diff
@@ -0,0 +1,126 @@
+func (client *TaskV2ClientImpl) GetTask(taskId string) (TaskV2, error) {
+	getTaskV2Url := client.getBaseUrl() + fmt.Sprintf("/operation/%s/core/operations/%s", client.orgId, taskId)
+	req, _ := http.NewRequest(http.MethodGet, getTaskV2Url, nil)
+	req.Header.Add(authnHeader, client.accessToken)
```
This repeats like 8 times.  It may make sense to have small struct for posting//getting http things done i.e. add the authn.

karaatanassov reviewed (commented):
Looks good. 

It may make sense to 

1. Use golang linter. 
2. Add configurable HTTP client. Do we need "insecure" flag for example?
3. Extract some "transport" abstraction that adds the token, may be formats urls and possibly handles common errors, unmarshalls JSON. I see similar code repeating 7-8 times.

Commit:
Added additional info for SDDC Groups in the schema

Additional info for SDDC Group queried from backend and placed in the resource schema for SDDC Group for user's convenience.

The additional info requires a second request and is for read-only data.

Included the "revive" linter and fixed issues.

Testing done:
make build
make test
golangci-lint run

Added a new E2E test for SDDC Groups
make testacc TESTARGS="-run=TestAccResourceSddcGroupZerocloud" - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

karaatanassov reviewed (approved):
Thank you for adding the linter and fixing all the code.
This looks good to me. Thank you for helping the Terraform users.
I do not understand deeply VMC so I trust you have done sufficient research on the necessary properties.

Commit:
Add documentation and examples for SDDC Group

Each new resource has a set of examples and documentation.

Testing done:
make build
make test
golangci-lint run

Added a new E2E test for SDDC Groups
make testacc TESTARGS="-run=TestAccResourceSddcGroupZerocloud" - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Add SDDC Group E2E test to the set of Zerocloud Lite E2E tests

Testing done:
make build
make test
golangci-lint run

Added a new E2E test for SDDC Groups
make testacc TESTARGS="-run=TestAccResourceSddcGroupZerocloud" - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>
